libevmasm/Assembly.cpp

```diff
@@ -297,6 +503,142 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
 	return root;
 }
 
+std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSON(
+	Json::Value const& _json,
+	std::vector<std::string> const& _sourceList,
+	size_t _level
+)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static std::set<std::string> const validMembers{".code", ".data", ".auxdata", "sourceList"};
+	for (std::string const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (Json::Value const& sourceName: _json["sourceList"])
+				solRequire(sourceName.isString(), AssemblyImportException, "The 'sourceList' array contains an item that is not a string.");
+		}
+	}
+	else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Member 'sourceList' may only be present in the root JSON object."
+		);
+
+	auto result = std::make_shared<Assembly>(EVMVersion{}, _level == 0 /* _creation */, "" /* _name */);
+	std::vector<std::string> parsedSourceList;
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		solAssert(_sourceList.empty());
+		for (Json::Value const& sourceName: _json["sourceList"])
+		{
+			solRequire(
+				std::find(parsedSourceList.begin(), parsedSourceList.end(), sourceName.asString()) == parsedSourceList.end(),
+				AssemblyImportException,
+				"Items in 'sourceList' array are not unique."
+			);
+			parsedSourceList.emplace_back(sourceName.asString());
+		}
+	}
+
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Member '.code' is missing.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Member '.code' is not an array.");
+	for (Json::Value const& codeItem: _json[".code"])
+		solRequire(codeItem.isObject(), AssemblyImportException, "The '.code' array contains an item that is not an object.");
+
+	result->importAssemblyItemsFromJSON(_json[".code"], _level == 0 ? parsedSourceList : _sourceList);
+
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not a string.");
+		result->m_auxiliaryData = fromHex(_json[".auxdata"].asString());
+		solRequire(!result->m_auxiliaryData.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solAssert(dataIter.key().isString());
+			std::string dataItemID = dataIter.key().asString();
+			Json::Value const& dataItem = data[dataItemID];
```

Comment:
This is a bit convoluted - can't we just get `dataItem` directly from `dataIter`?

Reply:
I thought the same thing and apparently we can't. Though I only briefly looked at how the iterator is defined. I didn't see a  member for it, but maybe you can find some mechanism for it if you look closer. In any case, I gave up myself on this because there were too many other things to adjust here.

Reply:
Dereferencing the iterator should work, no?

Reply:
Maybe. Feel free to check and suggest a variant that will work :)

Reply:
Looks like it works. In any case, regarding your latest comment about priorities - I wouldn't worry about this too much, and I'd prefer not to push to the branch in case @r0qs is actively working on (which I'm assuming he is). So we can leave it for another time, unless of course, he notices this comment and does it himself :)

Reply:
Pushing should not be a problem as long as you use `--force-with-lease` and only push new commits without modifying any existing ones. Then for @r0qs it would be a simple `git rebase -i origin/import-asm-json-updated` (and some conflict resolutions, but everyone should be used to resolving conflicts by now :)).